src/unitxt/operators.py

```diff
@@ -213,6 +213,103 @@ def process(
+    been processed by RenderInstruction StreamInstanceOperator (from renderer.py) and hence to contain a non None
+    field "instruction".
+    Demos are assumed to be included as well, in a field whose name is specified by argument 'demos_field'. That
+    field is assumed to contain a list of dicts, each containing keys: "source" and "target" of a single demo.
+    And finally, system_prompt is also assumed to be included
+    in the instance, in a field whose name is specified by argument 'system_prompt'.
+    The value in field "source" of the input instance describes (verbalizes) original values in the instance (as read
+    from the source dataset), in the context of the underlying task.
+    WholeInputFormatter makes this (verbalization of the) input into a whole-input-to-the-model, by combining it
+    with instruction, demos, and system_prompt, read from the input instance, through both formatting args
+    formatting arguments.
+    Following the Renderers action, the formatted model input is fed into field "source" (overwriting the field)
+    Also following renderers, this operator pops fields "instruction" and "demos" from the input instance.
+
+    Args:
+        demos_field (str): the name of the field that contains the demos, being dicts with "source" and "target" keys
+        system_prompt (str): the name of the field containing an introductory text and or xml tags
+        demo_format (str): formatting string for a single demo, combining fields source and target
+        model_input_format (str) overall output format, combining system_prompt, instruction, demos, and source
+    """
+
+    demos_field: str = "demos"
+    system_prompt: str = None
+    demo_format: str = (
+        "{source}\n{target}\n\n"  #  example: "User: {source}\nAgent: {target}\n\n"
+    )
+    model_input_format: str = (
+        "<SYS>{system_prompt}</SYS>\n{instruction}\n{demos}\n{source}\n"
+    )
+    instruction_prefix: str = ""  # potential prepend for non "" instruction. necessary for backward compatibility with iclformat
+
+    @staticmethod
+    def _retrieve_field_and_assert_not_none(instance, field_name) -> str:
+        if field_name is not None and field_name in instance:
+            field_value = instance[field_name]
+            assert (
+                field_value is not None
+            ), f"Value in field '{field_name}' should not be none. Received instance: {instance}"
+            return field_value
+        return ""
+
+    def process(
+        self, instance: Dict[str, Any], stream_name: Optional[str] = None
+    ) -> Dict[str, Any]:
+        assert (
+            "source" in instance
+        ), f"field 'source' is expected to be in the input instance. Received instance: {instance}"
+        source = self._retrieve_field_and_assert_not_none(
+            instance=instance, field_name="source"
+        )
+
+        system_prompt = self._retrieve_field_and_assert_not_none(
```

Comment:
Repeating my comment above, why are `system_prompt` and `instruction` fields of the instance and not explicit parts of the `model_input_format` parameter?

Reply:
This is, as far as I understand, because this operator is expected to follow other operators, one of which prepares the instruction, another one prepares the template (btw, Elron plans to combine these two into one).